### ISSUE TITLE
What's the recommended path to get acq yaml/reporting on manual catalog?

### ISSUE THREAD

**Issue by jeanconn** [sot/proseco#50]

**Comment by taldcroft** [comment]
Would it be feasible to have Matlab call an `AcqTable` method to do a manual edit, or more likely update the catalog after a series of edits which are confirmed?

This gets to idea that has been percolating that the backend Python session is persistent, so the `AcqTable` objects for all the obsids in the schedule can likewise be persistent and can be manipulated thusly.

So if there is just a method saying effectively "here is the new catalog after manual edits", then that method can handle the implementation details of ensuring that the YAML is consistent.

**Comment by taldcroft** [comment]
This persistence is probably a practical requirement if you want to allow things like a button to tell it to create the report for a particular obsid.

**Comment by jeanconn** [comment]
I thought when we were talking about running the calculations on the non-optimized catalog that there was already a method to just pass the needed pieces back to get this calculation.  Wasn't sure if I was just missing it.

I think I can pretty easily determine if the catalog is a 'saved' catalog or an automatically generated catalog, and I thought that 'saved'/manual might be enough (not if the the acq catalogs were changed, just if the final catalog had been saved after editing with any kind of change).

I wasn't planning to use persistent data in the Python session, but could work from that angle if need be.  In my first go, I've just saved the yaml text into the obsid starResults object.

**Comment by jeanconn** [comment]
And I'm still editing with the prospect of not making any buttons.

**Comment by taldcroft** [comment]
Generating the YAML is quite slow so it should be done only if required.  (It's slow because of all the flattening of table and numpy objects into pure Python).

**Comment by taldcroft** [comment]
*Eventually* there will be buttons.  BTW, on my old slow iMac:
```
In [3]: time acqs = acq.AcqTable.get_acq_catalog(19387)
CPU times: user 1.24 s, sys: 33.7 ms, total: 1.27 s
Wall time: 1.33 s

In [4]: time out = acqs.to_yaml()
CPU times: user 2.41 s, sys: 28.2 ms, total: 2.44 s
Wall time: 2.44 s
```

**Comment by jeanconn** [comment]
OK, so we will need to work more on the best ways to optimize.  

In my thinking, if they run star selection once with YAML saving and then we cache it on the Matlab side so the catalogs and YAML don't change unless the attitudes change (which I think is basically already done), there won't be much rework or additional slowness. 

I didn't see any way to get the YAML without making the YAML.

**Comment by taldcroft** [comment]
The point is to be lazy and only make the YAML on demand when you need it.  Which is to say do that along with some final export process that might already be taking some time.

**Comment by jeanconn** [comment]
I get you, it just doesn't plug in as neatly to what's already here (which is a model of "calculate stuff about a star catalog and then pass it around in the a data structure").  So different kinds of "lazy".

But if we've got a reliable cache of the values we want in the persistent session on Python side I can come around to that way of thinking.

**Comment by jeanconn** [comment]
So instead of needing a method to pass the acq catalog back to calculate probabilities, we'll need both that and some conventions for working with the cache of AcqTables.

**Comment by jeanconn** [comment]
And I'm still not seeing a reason for buttons, just as there's no "write out guide summary" button.

**Comment by taldcroft** [comment]
There needs to be some way for Eric to tell SAUSAGE to create the detailed HTML report.   He expressed that as a desirement at some point.

**Comment by jeanconn** [comment]
Right, I just thought I'd export the per-obsid yaml when the tool exports a dot.  I haven't looked yet to see what else needs to run to create the detailed HTML report.

**Comment by jeanconn** [comment]
It is also probably easy to add a type "proseco html" to the export menu.  I just specifically think I can get away without *buttons* just now :smile:

**Comment by taldcroft** [comment]
The convention for working with the cache is pretty easy, I think.
```
pyexec("catalogs[$obsid].do_something($data1, $data2)")
```
But maybe we need to slow down and think about this some more in terms of what is really necessary (vs what is possible).  Mark warned about not getting too crazy, and I agree.

**Comment by taldcroft** [comment]
Along the lines of my last comment, for instance, I don't think that creating the HTML reports in Matlab is a *requirement* for this release.

**Comment by jeanconn** [comment]
OK.  Though I thought the YAML was required and that's the focus.  Regarding the cache, the convention I meant to settle on was the utility that would be needed to check that acq catalogs match what was solved.

**Comment by taldcroft** [comment]
And just thinking now I am wondering about the persistent Python.  That might actually not be the case (e.g. shut down Matlab and pick up again).  They must have some persistence model for schedule building which doesn't depend on the CPU process persisting.

**Comment by taldcroft** [comment]
Yes, YAML is required. HTML is not.

**Comment by taldcroft** [comment]
That could work.  We can always unpack and save as YAML later in starcheck (to avoid issues with the pickle not being readable in Python 4, etc).

**Comment by jeanconn** [comment]
The bigger issue I think is whenever you have "separate" data that is not captured in the DOT, you've got potential for drift between the data structures.  So for star catalogs, I think if we, say, had B products just to do a "DOT" edit on a star catalog, would you want to try to load the previous yaml or pickle files from A products?  That starts to get tricky.  I think for first round here, we'd just plug all the B products star catalogs back into proseco for calculations.

Overall, I'm still wondering it *standard* process should just be to write out the yaml for all the catalogs as the non-optimized, manual versions plugged back into proseco for probability calcs. 

A separate process would be to write out the "as-solved YAML" for more intensive ACA review.

**Comment by taldcroft** [comment]
This has been a productive and illuminating discussion, and my thinking definitely evolved.

I think the guiding principle right now should be to maintain the current process within Matlab to the extent possible.  Here is a concept which you can comment on.

- SAUSAGE (via proseco) generates all the catalogs and stores this within its `starResults` object like now, where the only new feature is a new attribute which is a binary blob of the pickled proseco info.   I'm favoring pickle now because it is 100 times faster and 3 times smaller, and fine for internal use.
- I'm assuming that the `starResults` objects are persisted somehow between sessions, so the proseco info pickle persists.  This is important.
- At the time of DOT export the pickles get written.  In fact this could be to a single file by pickling a structure which is a dict of the info pickles keyed by obsid.  But these details are TBD for sure.
- For hand edits, no calls back to proseco or mods of the pickled proseco objects is done.  Those info pickles represent the proseco suggested catalog and how it got there.
- Starcheck will check the consistency of the backstop catalog with the proseco-suggested catalog and issue an orange warning if they are different.  The idea is really that hand edits will be very rare.   Orange (not blue or yellow) because it *could* be a real problem. Reviewers will be aware and interpret the proseco report accordingly.  
- This is the concept for the 2018-Oct release with proseco 4.0.  We could refine as experience dictates, but for the initial rollout this has fewer moving parts.

Between A and B products, again the "do whatever they are doing now" principle holds.  I don't know if they restore all the starResults from a Matlab save file, or if they regenerate the catalogs for a new product set.  I have a vague memory that it might be the latter, but in either case we are set.  The proseco info pickles will be available for export at the time of DOT export.

**Comment by jeanconn** [comment]
> I'm assuming that the starResults objects are persisted somehow between sessions, so the proseco info pickle persists. This is important.

You've declared it important, but I don't completely understand.  I think there is just one session.

**Comment by jeanconn** [comment]
My understanding is that the way this would work would be that planner A would save a mat file of his ORViewer schedule session (no catalogs made yet) and planner B would load up that schedule in a new, get to the end, and do the final run through SAUSAGE that does the catalog replacement.  SAUSAGE itself would run only in that second session.

**Comment by jeanconn** [comment]
> Between A and B products, again the "do whatever they are doing now" principle holds. I don't know if they restore all the starResults from a Matlab save file, or if they regenerate the catalogs for a new product set. I have a vague memory that it might be the latter, but in either case we are set. The proseco info pickles will be available for export at the time of DOT export.

I think "we are set" is not quite right.  A little more to refine.  For A to B products just for a star edit (no schedule edits), I don't think they load the schedule mat file with the previous starResults, I think they just load the DOT in SAUSAGE.  The proseco pickle in starResults will not be available in B products unless we rerun proseco on the catalogs.  That would wipe out any previous manual edits, which is expliticitly *not* what we want.  So this needs more thinking.  We could use the saved pickle file that we'd already made, or have an another run of proseco on all the catalogs that would just save out what the optimized proseco catalog would look like for that attitude but not overwrite anything in there.

In fact, that might be the safest thing anyway ... to have the final save-the-DOT process save the actual catalogs (manually edited or saved from any previous proseco calls), but add a seprate full rerun of proseco acq on all the catalogs and save a pickle of "what proseco *would* have selected for this catalog".  This does increase the time at the very end for the planners, but I think that would be doable.

**Comment by jeanconn** [comment]
> The proseco pickle in starResults will not be available in B products unless we rerun proseco on the catalogs.

Though actually, I'll think about this more in the context of what they do now.  For example, if you load a DOT in SAUSAGE that already has updated star fields, even if you don't replace a catalog I *think* what they do now is they rerun SAUSAGE star selection when you get to the obsid/field, but only keep the "side effects" of the process (starResult error flags and such).  You were hinting at this as well.  

While it would work to use proseco that way to regenerate the proseco data (without replacing catalogs), that would be much slower. 

> In fact, that might be the safest thing anyway ... to have the final save-the-DOT process save the actual catalogs (manually edited or saved from any previous proseco calls), but add a separate full rerun of proseco acq on all the catalogs and save a pickle of "what proseco *would* have selected for this catalog".  This does increase the time at the very end for the planners, but I think that would be doable.

And rethinking this, I think it will be perfect fine to do something like "check each observation to see if it has proseco pickle data, run proseco in non-catalog-update mode for any that don't, write out the pickle to a file".  That makes your final save for the simple B product star catalog edit take a little longer, but keeps the nominal A products process as short as possible.

**Comment by taldcroft** [comment]
Can we learn anything from the process and code for the guide star summary?  It seems somewhat analogous.  For B products with a catalog edit, it somehow manages to know and output the selection stage of every star in every catalog, including the ones that were hand-edited.

**Comment by taldcroft** [comment]
BTW, I think we are at the point where you have good questions and we would benefit from input from the FOT.

**Comment by jeanconn** [comment]
I *think* that's the way that it reruns SAUSGE star selection on all observations and keeps the "side-effects" (stage selection would be in there too).   

But I still think using proseco exactly that way would be slow.

**Comment by taldcroft** [comment]
FYI proseco is about as fast as SAUSAGE, I think.  Call it 0.5 sec per obsid.

**Comment by taldcroft** [comment]
Though it will slow down in cases with problematic fid lights / acq star interactions.

**Comment by jeanconn** [comment]
> But I still think using proseco exactly that way would be slow.

I should have said, would be slow and would have no point.  

Still considering the B products just-star catalog edit.  You've loaded the DOT in SAUSAGE and you go to one catalog and make an edit.  Then you save/export the DOT.

So, proseco is going to need to run on all the unproseco-d catalogs at the export DOT stage to remake the pickles.  Will it really matter at that point if we already have a proseco pickle for the one field that was edited?  

But I'll think about all of this as I prototype and develop more questions for FOT.

**Comment by jeanconn** [comment]
And another thought for the thread: earlier there was the idea of just asking for all of the proseco info/yaml/pickl data from the persistent Python session.  I think that could also still work.  I think @taldcroft dropped that idea with the thought that saving the pickles in the starResults would be persistent across sessions... but I don't think those starResults are really saved to mat files in regular process so they don't get that "win" of being persistent that way anyway.

**Comment by jeanconn** [comment]
Looking back at this days later, I'm not sure we settled all parts of the original question.  Specifically for a manual catalog, did you want to leave starcheck's probability calculations alone or is it supposed to call some proseco methods to calculate and use the marginalized probabilities?

**Comment by taldcroft** [comment]
It'll need to call proseco to get the right answer.  We'll figure out how to make that happen a bit further down the line.

**Comment by jeanconn** [comment]
If we have the idea to pass the pickle back across into matlab I'm not quite sure how to do that with the pickle bytes... If I try to decode() the pickle data before passing, even using pickle protocol 0, I seem to end up with errors.

**Comment by taldcroft** [comment]
With #69 you can generate an AcqTable using fully-specified `include_ids` and `include_halfws` for the manual catalog.  This gives access to probabilities and a report that is relevant to that catalog.  When there are one or more `include_ids` then those are locked out from optimization.  If all the slots are specified, then no optimization calls are made, but all the machinery for candidate acquisition stars is still run, so the report can still reflect that info and show the relevant images.

- In starcheck this can be used for a catalog that does not match the catalog in the pickle.  But maybe that would not happen if we do the Matlab process in the next bullet.
- In Matlab this can be used for a catalog is that has been edited.  So maybe in the end it makes sense to do this (specify all selected stars in the case of a hand-edit) instead of putting the original optimization run into the pickle.

**Comment by jeanconn** [comment]
> In Matlab this can be used for a catalog is that has been edited. So maybe in the end it makes sense to do this (specify all selected stars in the case of a hand-edit) instead of putting the original optimization run into the pickle.

So for that, it does introduce the issue that I'm not explicitly tracking *if* the catalog has been modified which we'd need to hash that out a bit more for interface.

**Comment by taldcroft** [comment]
The code must track that *somewhere* since it is evident in the guide summary, right?

**Comment by jeanconn** [comment]
Right. I *think* manually selected items are marked as being "selected in stage 0" and then that gets converted to an X for the maneuver summary.

https://github.com/sot/fot_matlab_tools/blob/master/MissionScheduling/stars/%40ACACatalog/private/mgSummary.m#L73

Since I've faked in our selections as stage 1 I think this will still work.  So we should have some tracking, just not sure what we'd need for functions/interface here?  Look for changes to the fid or acq stars selected at stage 0 and pass the edited catalog to the new code to get updated probabilities?  

I note in passing that I don't think we going to see if a star has been deleted.  I don't see that as a big deal, but something we should probably add a check for in starcheck at some point (that the commanded catalog matches the catalog used for probabilities).

**Comment by taldcroft** [comment]
This sounds fine.  I would say that we do not have a *requirement* to get this in for 4.0.  

Long ago before we figured out how we could do this, there was the idea of just tossing in the catalog metadata (log and probs etc) from the catalog proseco *would* have chosen.  Then starcheck can deal with tidying things up by carefully detecting if there are diffs (including just a deleted star).  

Given time pressure, this seems like a reasonable path.  Thoughts?

**Comment by jeanconn** [comment]
Sounds good. Regarding interface, is your idea that get_aca_catalog will just have include_ids and  include_halfws in the keyword args (which will be filled if logic on the matlab side says that the catalog has been modified)?  Do we need any other keywords or attributes for this?

**Comment by taldcroft** [comment]
> Regarding interface, is your idea that get_aca_catalog will just have include_ids and include_halfws in the keyword args (which will be filled if logic on the matlab side says that the catalog has been modified)? 

Yes.  I'm not aware of additional kwargs that are required.

**Comment by jeanconn** [comment]
@taldcroft Regarding the output pickle from the matlab tools, with which protocol should that be written?

**Comment by taldcroft** [comment]
> Do we need to specify a pickle protocol when writing out the files?

I don't see a problem using the default (highest) protocol.  The only issue with using the default is compatibility with older versions of Python (e.g. 2.7).  From now on we will only be using a Python that is newer than the one which wrote a proseco pickle.  

A much bigger problem with pickle is that the proseco classes might change enough to be incompatible.  We'll cross that bridge if it ever becomes a problem.
